kms_log/main.tf

```diff
@@ -0,0 +1,398 @@
+    kms_data_key_reuse_period_seconds = 600
+    message_retention_seconds = 604800 # 7 days
+    tags = {
+        environment = "${var.env_name}"
+    }
+}
+
+resource "aws_sqs_queue" "kms_ct_events" {
+    name = "${var.env_name}-kms-ct-events"
+    delay_seconds = 60
+    max_message_size = 2048
+    visibility_timeout_seconds = 60
+    message_retention_seconds = 345600 # 4 days
+    kms_master_key_id = "${aws_kms_key.kms_logging.arn}"
+    kms_data_key_reuse_period_seconds = 600
+    redrive_policy = <<POLICY
+{
+    "deadLetterTargetArn": "${aws_sqs_queue.dead_letter.arn}",
+    "maxReceiveCount": 10
+}
+POLICY
+tags = {
+        environment = "${var.env_name}"
+    }
+}
+
+resource "aws_sqs_queue_policy" "default"{
+    queue_url = "${aws_sqs_queue.kms_ct_events.id}"
+    policy = "${data.aws_iam_policy_document.sqs_kms_ct_events_policy.json}"
+}
+
+data "aws_iam_policy_document" "sqs_kms_ct_events_policy" {
+    statement {
+        sid = "Allow CloudWatch Events"
+        effect = "Allow"
+        actions = ["sqs:SendMessage"]
+        principals {
+            type = "Service"
+            identifiers = ["events.amazonaws.com"]
+        }
+        resources = ["${aws_sqs_queue.kms_ct_events.arn}"]
+        condition {
+            test = "StringLike"
+            variable = "aws:SourceArn"
+            values = [
+                "${aws_cloudwatch_event_rule.decrypt.arn}"
+            ]
+
+        }
+    }
+}
+
+# cloudwatch event rule to capture decryption events
+resource "aws_cloudwatch_event_rule" "decrypt" {
+    count = "${var.kmslogging_service_enabled}"
+    name = "${var.env_name}-decryption-events"
+    description = "Capture decryption events"
+
+    event_pattern = <<PATTERN
+{
+    "source": [
+        "aws.kms"
+    ],
+    "detail-type": [
+        "AWS API Call via CloudTrail"
+    ],
+    "detail": {
+        "eventSource": [
+            "kms.amazonaws.com"
+        ],
+        "requestParameters": {
+            "encryptionContext": {
+                "context": [
+                    "password-digest",
+                    "pii-encryption"
+                ]
+            }
+        },
+        "resources": {
+            "ARN": [
+                "${data.aws_kms_key.application.arn}"
+            ]
+        },
+        "eventName": [
+            "Decrypt"
+        ]
+    }
+}
+PATTERN
+}
+
+resource "aws_cloudwatch_event_target" "sqs" {
+    rule = "${aws_cloudwatch_event_rule.decrypt.name}"
+    target_id = "${var.env_name}-sqs"
+    arn = "${aws_sqs_queue.kms_ct_events.arn}"
+}
+
+# dynamodb table for event correlation
+resource "aws_dynamodb_table" "kms_events" {
+    name = "${local.dynamodb_table_name}"
+    billing_mode = "PAY_PER_REQUEST"
+    hash_key = "UUID"
+    range_key = "Timestamp"
+
+    attribute {
+        name = "UUID"
+        type = "S"
+    }
+
+    attribute {
+        name = "Timestamp"
+        type = "S"
+    }
+
+    attribute {
+        name = "Correlated"
+        type = "S"
```

Comment:
TODO: check against latest AWS provider and file bug against terraform-provider-aws if `BOOL` is not supported